packages/maskbook/src/plugins/ITO/hooks/usePoolPayload.ts

```diff
@@ -21,6 +21,6 @@ async function suspender(pid: string) {
     try {
         storage.set(pid, await PluginITO_RPC.getPool(pid))
     } catch (error) {
-        storage.set(pid, {} as JSON_PayloadInMask)
+        storage.set(pid, new Error())
```

Comment:
a bit strange why don't you just throw an error here?

Reply:
It will cause a dead loop which keep throwing error, and the UI is keeping ITO loading if throw here.

Reply:
I don't clearly understand why it would happen. If the second render goes normal, it will have no error and goes to line 22 right?

Reply:
Yes, if `await PluginITO_RPC.getPool(pid)` works well.

Reply:
you mean that screen video, I make the error intentionally 

```ts
i = 0

async function suspender(pid: string) {
  try {
    if (i === 0) {
      i++
      throw new Error()
    }
    storage.set(pid, await PluginITO_RPC.getPool(pid))
  } catch {
    storage.set(pid, new Error())
  }
}
```

Reply:
Ok, this makes sense I understand why now. https://github.com/facebook/react/issues/17526

Reply:
So according to the link above, to make things correct, we should do like this:

```js
const cache = storage.get(...)
// cache match
if (isError(cache)) throw cache // Must sync
else return cache

throw getPool(pid)
```

But I got no idea how to clear the error cache when recovering from the ErrorBoundary

Reply:
Maybe like this: Only throw once then clear the error one? Or clear the whole cache?

```js
if (isError(cache)) { storage.remove(pid); throw cache }
```

Reply:
The problem here is that the storage is global so it won't be cleared out in the error recovery case

Reply:
Ok I think this can resolve the problem:

```js
const storage = new Map()
function useSomething(id) {
    const [fatalError, setError] = useState(null)
    const result = storage.get(id)
    if (result) return result
    if (fatalError) throw fatalError
    if (!result) throw suspender(id).catch(setError)
}
```

Storage error in this way, we can automatically clear the error when the component tree is recovered from the ErrorBoundary. @zhouhanseng would you like to have a try?

Reply:
```ts
if (isError(cache)) { storage.remove(pid); throw cache }
```

I don't quite follow from those code snippets.

Reply:
@Jack-Works sure

Reply:
@Jack-Works The code below doesn't work as expection, 
Expect: it would display the <ITO_Failue />
Actual: always show <ITO_Loading />

```ts
import { PluginITO_RPC } from '../messages'
import type { JSON_PayloadInMask } from '../types'
import { useState } from 'react'

const storage = new Map<string, JSON_PayloadInMask | Error>()

export function usePoolPayload(pid: string) {
    const [fatalError, setError] = useState(null)
    const result = storage.get(pid)

    if (fatalError) throw fatalError

    if (!result) {
        throw suspender(pid).catch(setError)
    }

    return {
        payload: storage.get(pid)!,
        retry: () => retry(pid),
    }
}

async function retry(pid: string) {
    storage.delete(pid)
    if (!storage.has(pid)) throw suspender(pid)
    return
}

async function suspender(pid: string) {
    try {
        // if (i === 100) {
            // i++
        throw new Error()
        // }
        // storage.set(pid, await PluginITO_RPC.getPool(pid))
    } catch (error) {
        throw error
        // storage.set(pid, new Error())
    }
}

```

Reply:
```ts
import { PluginITO_RPC } from '../messages'
import type { JSON_PayloadInMask } from '../types'
import { useState } from 'react'

const storage = new Map<string, JSON_PayloadInMask>()

export function usePoolPayload(pid: string) {
    const [fatalError, setError] = useState(null)
    const result = storage.get(pid)

    if (result) return {
        payload: storage.get(pid)!,
        retry: () => retry(pid),
    }
    if (fatalError) throw fatalError
    throw suspender(pid).catch(setError)
}

async function retry(pid: string) {
    storage.delete(pid)
    throw suspender(pid)
}

async function suspender(pid: string) {
    return storage.set(pid, await PluginITO_RPC.getPool(pid))
}
```

Try this one @zhouhanseng

Reply:
@Jack-Works 

```ts
async function suspender(pid: string) {
    throw new Error()
}
```

Still always show <ITO_Loading />

Reply:
@zhouhanseng please tell me where can I invoke this hook, I will look at this later today

Reply:
https://twitter.com/Zhouhancheng/status/1352487007411269632